\section{Introduction} 
\label{sec:introduction}

Next-generation sequencing (NGS) technologies have attracted a large amount of attention from both researchers and clinicians. 
Human DNA samples are chopped into billions of small fragments, called reads, and a sequencer determines the order of each read's nucleotides. 
A software aligner then maps the billions of sequenced reads onto a reference human genome, which entails tremendous computational challenges.
%%%The size of a DNA sequence is often measured in base pairs (bps). 
A read typically consists of hundreds of nucleotides or base pairs (bps), while the reference genome contains 3.2 billion bps \cite{Mardis2008}.
%%The advent of next-generation sequencing (NGS) technologies dramatically reduces the cost of genome sequencing. 
%%Today's sequencing technologies can obtain a genome for an individual for \$1,000 or less \cite{Mardis2006}. 
%%The technology can be widely used in research and is transitioning into the clinic for applications such as precision medicine for cancer treatment. 
%%Next-generation sequencers are more cost-effective because they generate the sequence 
%%from very small fragments (reads) of length in the range of a few hundred nucleotides. 
%%Mapping billions of sequenced fragments onto a genome sequence by taking advantage of the known human genome sequence 
%%is called resequencing, and it entails tremendous computational challenges.

State-of-the-art read aligners, such as BWA-MEM \cite{BWA-MEM} and Bowtie 2 \cite{Bowtie2}, 
carry out the read mapping in two steps. 
First, each read is fragmented into small pieces called seeds, and mapped to the reference genome. 
The mapping of seeds to the reference genome must be exact, i.e., no gap or mismatch is allowed. 
A backward search algorithm based on the Burrows-Wheeler Transform (BWT) \cite{BWT} takes only O($m$) time for a seed of length $m$ to map to the super-long reference genome, independent of the size of the reference genome, and therefore is employed by almost all contemporary sequence aligners.

In the second step, each seed map gets extended leftward and rightward to span the entire read. 
The extensions allow inexact mappings, in which gaps and mismatches are allowed. A predefined scoring function is provided for evaluating the effectiveness of inexact mappings. Only the ones achieving high enough scores are recorded in the output. 
The Smith-Waterman (S-W) algorithm, a dynamic programming (DP) algorithm with quadratic time complexity, is a commonly used approach to address this problem. 
It is the main computation bottleneck in state-of-the-art aligners like BWA-MEM (30\%-50\% computation time) \cite{BWA-MEM}. 
In this work we focus on accelerating the S-W algorithm in BWA-MEM. The methodology, however, can be applied to other contemporary aligners, such as Bowtie 2 \cite{Bowtie2} and LAST \cite{LAST}. 

The S-W algorithm is inherently anti-diagonal parallelizable, since the elements along the anti-diagonal direction in the 2D table used for the DP procedure are independent of each other \cite{Edmiston1988}. 
This feature, called wavefront, has been explored in different ways on different platforms \cite{Preusser2012}\cite{RaceLogic}\cite{Zhang2007}\cite{Lam2013}. 
It works fairly well when input sizes are homogeneous. 
However, the solutions that target the general S-W algorithm do not fit well for the optimized version used in BWA-MEM. 
First, a huge number of reads at the billion scale need to be processed with high throughput. 
Conventional approaches overemphasize the inner-task parallelism while neglecting the task-level parallelism. 
Second, conventional wavefront-based architectures cannot utilize computational resources efficiently when the input sizes vary sharply. 
Third, the pruning heuristic used in BWA-MEM prevents conventional solutions from being adopted directly and efficiently.

In this paper, we propose an architecture to address these issues.
The novelties of our architecture can be summarized as follows: 
(1) we propose an array-based architecture for processing the enormous number of reads in a high-throughput fashion, adapting better to inputs with widely varied sizes; 
(2) we provide a two-level hierarchical architecture for resource management, avoiding wasting resources in synthesizing bus interfaces while satisfying the off-chip bandwidth demand;
(3) our design supports the pruning technique, shortening the runtime of S-W significantly. 

Our FPGA implementation demonstrates a 26.4x speedup compared to a 24-thread Intel Haswell Xeon server for the S-W algorithm in BWA-MEM. 
We also show that our design outperforms wavefront-based S-W implementations by up to 6x with the same FPGA resource utilization.
